src/readblock.c

```diff
@@ -0,0 +1,299 @@
+/****************************************************************************************
+
+    FFIO - Flexible format I/O
+
+    FFIO provides functionality for reading proj style instruction strings written
+    in a less strict format than usual:
+
+    *  Whitespace is generally allowed everywhere
+    *  Comments can be written inline, '#' style
+    *  ... or as free format blocks
+
+    The overall mission of FFIO is to facilitate communications of geodetic
+    parameters and test material in a format that is highly human readable,
+    and provides ample room for comment, documentation, and test material.
+
+    See the PROJ ".gie" test suites for examples of supported formatting.
+
+****************************************************************************************/
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdarg.h>
+
+#include <string.h>
+#include <ctype.h>
+
+#include <math.h>
+#include <errno.h>
+
+
+
+static const char *gie_tags[] = {
+    "<gie>", "operation", "accept", "expect", "roundtrip", "banner", "verbose",
+    "direction", "tolerance", "builtins", "echo", "</gie>"
+};
+
+static const size_t n_gie_tags = sizeof gie_tags / sizeof gie_tags[0];
+
+typedef struct {
+    FILE *f;
+    const char **tags;
+    const char *tag;
+    char *args;
+    char *next_args;
+    size_t n_tags;
+    size_t args_size;
+    size_t next_args_size;
+    size_t argc;
+    size_t lineno, next_lineno;
+    size_t level;
+}  ffio;
+
+
+/***************************************************************************************/
+static ffio *ffio_create (const char **tags, size_t n_tags, size_t max_record_size) {
+/****************************************************************************************
+
+****************************************************************************************/
+    ffio *F = calloc (1, sizeof (ffio));
+    if (0==F)
+        return 0;
+
+    if (0==max_record_size)
+        max_record_size = 1000;
+
+    F->args = calloc (1, 5*max_record_size);
+    if (0==F->args) {
+        free (F);
+        return 0;
+    }
+
+    F->next_args = calloc (1, max_record_size);
+    if (0==F->args) {
+        free (F->args);
+        free (F);
+        return 0;
+    }
+
+    F->args_size = 5*max_record_size;
+    F->next_args_size = max_record_size;
+
+    F->tags = tags;
+    F->n_tags = n_tags;
+    return F;
+}
+
+
+/***************************************************************************************/
+static ffio *ffio_destroy (ffio *F) {
+/****************************************************************************************
+
+****************************************************************************************/
```

Comment:
What about `f`, `tags`, and `tag`?  If you don't need to handle them, add a comment as to why.